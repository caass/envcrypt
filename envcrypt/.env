SOME_ENVIRONMENT_VARIABLE="TO_MAKE_DOCTESTS_PASS"
