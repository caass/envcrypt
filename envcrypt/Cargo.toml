[package]
authors = ["Cass Fridkin <cass@swag.lgbt>"]
categories = ["cryptography", "development-tools::build-utils"]
description = "Drop-in replacement for env! that encrypts your variables at compile-time and decrypts them at runtime, preventing naughty folks from snooping your binary for secrets or credentials."
edition = "2021"
keywords = [
  "environment-variables",
  "encrypt-environment",
  "encrypt-enviroment-variables",
  "encrypt-env",
  "encrypt-env-vars",
]
license = "MIT OR Apache-2.0"
name = "envcrypt"
readme = "../Readme.md"
repository = "https://github.com/caass/envcrypt"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
envcrypt-macro = { path = "../envcrypt-macro" }
magic-crypt    = "3.1.10"
