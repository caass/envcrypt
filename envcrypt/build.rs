fn main() {
    println!("cargo:rustc-env=SECRET_KEY=SOME_FAKE_SECRET_KEY_FOR_DOCTESTS");
    println!("cargo:rustc-env=CLIENT_SECRET=ANOTHER_SECRET_FOR_DOCTESTS");
}
