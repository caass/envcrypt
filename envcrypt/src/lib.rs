#![deny(missing_docs, clippy::missing_docs_in_private_items)]
#![cfg_attr(windows, doc = include_str!("..\\README.md"))]
#![cfg_attr(not(windows), doc = include_str!("../README.md"))]

mod decrypt;

/// For internal use only.
#[doc(hidden)]
pub mod __internal {
    #[doc(hidden)]
    pub use super::decrypt::decrypt;
}

/// Inspects and encrypts an environment variable at compile time
/// and decrypts at runtime.
///
/// This macro will expand to the value of the named environment variable at compile time,
/// yielding an expression of type `&'static str`.
/// Use [`std::env::var`] instead if you want to read the value at runtime.
///
/// ```rust
/// # use envcrypt::envc;
/// let path: &'static str = envc!("PATH");
/// println!("the $PATH variable at the time of compiling was: {path}");
/// ```
///
///  If the environment variable is not defined, then a compilation error will be emitted.
/// To not emit a compile error, use the [`option_envc!`] macro instead.
///
/// ```compile_fail
/// # use envcrypt::envc;
/// let unlikely_variable: &'static str = envc!("HIGHLY_UNLIKELY_ENVIRONMENT_VARIABLE");
/// ```
#[doc(inline)]
pub use envcrypt_macro::envc;

/// Optionally inspects and encrypts an environment variable at compile time
/// and decrypts at runtime.
///
/// If the named environment variable is present at compile time,
/// this will expand into an expression of type `Option<&'static str>`
/// whose value is `Some` of the value of the environment variable.
/// If the environment variable is not present, then this will expand to `None`.
///
/// Use [`std::env::var`] instead if you want to read the value at runtime.
///
/// A compile time error is never emitted when using this macro
/// regardless of whether the environment variable is present or not.
///
/// ```rust
/// # use envcrypt::option_envc;
/// let key: Option<&'static str> = option_envc!("SECRET_KEY");
/// println!("the secret key might be: {key:?}");
/// ```
#[doc(inline)]
pub use envcrypt_macro::option_envc;
