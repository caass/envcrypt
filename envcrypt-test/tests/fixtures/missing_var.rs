use envcrypt::envc;

fn main() {
    envc!("MISSING")
}
