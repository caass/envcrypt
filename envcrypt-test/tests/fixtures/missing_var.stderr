error: environment variable `MISSING` not defined
 --> tests/fixtures/missing_var.rs:4:5
  |
4 |     envc!("MISSING")
  |     ^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `envc` (in Nightly builds, run with -Z macro-backtrace for more info)
